### PR TITLE
Implement redux storage

### LINKED ISSUE #22: Investigate the use of IndexedDB to store state

### PR DESCRIPTION
**Pull request by lalexgap**
This introduces `redux-storage` and stores channel info (and some related info) in storage.

We load from indexedDB when the page is loaded and we save to indexedDB after every action(this can be configured to a specific list of actions). Currently there's no error handling for storage related failures(#629).

Instead of storing `adjudicatorState` there is saga that updates `adjudicatorState` based on the latest on-chain info.

A quick way of seeing storage working is to
- Open a channel, play a game, and select `Keep Open` when concluding
- Refresh one more browsers
- Open a new channel. It will use the existing ledger channel that was saved in storage.

Fixes #22.

### CODE REVIEW

File: packages/wallet/src/redux/reducer.ts [hunk 1]
```diff
@@ -201,7 +201,13 @@ const waitForLoginReducer = (
 ): states.WalletState => {
   switch (action.type) {
     case 'WALLET.LOGGED_IN':
-      const { privateKey, address } = ethers.Wallet.createRandom();
+      let address = state.address;
+      let privateKey = state.privateKey;
+      if (!address || !privateKey) {
+        const { privateKey: newPrivateKey, address: newAddress } = ethers.Wallet.createRandom();
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/reducer.ts]:
```suggestion
        ({ privateKey, address } = ethers.Wallet.createRandom());
```

File: packages/wallet/src/redux/reducer.ts [hunk 2]
```diff
@@ -201,7 +201,13 @@ const waitForLoginReducer = (
...
+        address = newAddress;
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/reducer.ts]:
```suggestion
```

File: packages/wallet/src/redux/reducer.ts [hunk 3]
```diff
@@ -201,7 +201,13 @@ const waitForLoginReducer = (
...
+        privateKey = newPrivateKey;
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/reducer.ts]:
```suggestion
```

File: packages/wallet/src/redux/reducer.ts [hunk 4]
```diff
@@ -201,7 +201,13 @@ const waitForLoginReducer = (
 ): states.WalletState => {
   switch (action.type) {
     case 'WALLET.LOGGED_IN':
-      const { privateKey, address } = ethers.Wallet.createRandom();
+      let address = state.address;
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/reducer.ts]:
```suggestion
      let { address, privateKey } = state;
```

File: packages/wallet/src/redux/reducer.ts [hunk 5]
```diff
@@ -201,7 +201,13 @@ const waitForLoginReducer = (
...
+      let privateKey = state.privateKey;
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/reducer.ts]:
```suggestion
```

File: packages/wallet/src/redux/adjudicator-state/reducer.ts [hunk 6]
```diff
@@ -59,6 +61,14 @@ const fundingReceivedEventReducer = (
   const { channelId } = action;
   return setBalance(state, channelId, action.totalForDestination);
 };
+const channelUpdateReducer = (state: AdjudicatorState, action: actions.ChannelUpdate) => {
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/adjudicator-state/reducer.ts]: Is this related to storing the redux-store in storage?

File: packages/wallet/src/redux/state.ts [hunk 7]
```diff
@@ -56,6 +56,8 @@ export interface ChannelSubscriptions {
 }
 export interface WaitForLogin extends SharedData {
   type: typeof WAIT_FOR_LOGIN;
+  address?: string;
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/state.ts]: I'm wondering if it's better not to store the address, which is a pure function of the private key.

File: packages/wallet/src/redux/adjudicator-state/reducer.ts (hunk 6 above)
Comment (lalexgap) [reply packages/wallet/src/redux/adjudicator-state/reducer.ts]: Yes, it's used to update the `adjudicator-state` on load for stored channels. We currently don't store the `adjudicator-state` so we need to initialize it on load from storage.

File: packages/wallet/src/redux/state.ts (hunk 7 above)
Comment (lalexgap) [reply packages/wallet/src/redux/state.ts]: Going to leave this as is for now to get the PR in, but perhaps something we can consider in the future?

File: packages/wallet/src/redux/store.ts [hunk 8]
```diff
@@ -1,16 +1,33 @@
...
 
 import { walletReducer } from './reducer';
 import { sagaManager } from './sagas/saga-manager';
+import filter from 'redux-storage-decorator-filter';
+import createEngine from 'redux-storage-engine-indexed-db';
+
+// We currently whitelist the values that we store/load.
+const storageEngine = filter(createEngine('magmo-wallet'), [
+  'whitelisted-key',
+  ['address'],
```
Comment (andrewgordstewart) [inline packages/wallet/src/redux/store.ts]: A comment that didn't seem to make it into my review. This part is tightly coupled with the shape of the shape of the redux store. Perhaps the whitelisted keys should be defined next to that interface and imported here?
